headers/validation.h

```diff
@@ -260,15 +335,19 @@ struct GeometricDistributionGenerator {
 
 	double p_;
 
-	explicit GeometricDistributionGenerator(double p) : p_(p) {}
+	explicit GeometricDistributionGenerator(double p) : p_(p) {
+		assert(p_ > 0);
+		assert(p_ < 1);
+	}
 
 	// NOTE: Currently this retries instead of clamping to the interval.
 	T operator()(T low, T high, std::mt19937_64& rng) const {
+		assert(low <= high);
 		T v;
-		do {
-			v = low + std::geometric_distribution<T>(p_)(rng);
-		} while(v < low or v > high);
-		return v;
+		while (true) {
+			v = low + std::floor(std::log(Random::real64(rng)) / std::log1p(-p_));
```

Comment:
This could also use a source ;)
(I'm too lazy to think about formulas at the moment, and some indication for why this is correct would be nice.)

Reply:
https://en.wikipedia.org/wiki/Geometric_distribution
look for  "The exponential distribution is the continuous analogue of the geometric distribution[...]"